### ISSUE TITLE
process leak of rt helper. [JIRA: RIAK-2777]

### ISSUE THREAD

**Issue by ksauzz** [basho/riak_repl#649]
#### Summary

A customer hit memory leak problem of a sink cluster such that leak memory grows upto 10~20GB at #zd9123. Many riak_repl2_rtsink_helper processes were found in cluster info, which hog ~600MB per a process, and look orphaned.
Unstable network connectivity can cause this because many tcp_error were found in console.log when memory usage were increasing, and I reproduced this problem by simulating them.

You can find whole cluster_info and memory usage graphs at #zd9123. (Most text is in Japanese. sorry.)

In my opinion, to resolve it we should stop riak_repl2_rtsink_helper in riak_repl2_rtsink_conn:terminate/2 but I'm not sure its side effect. According to my test, also riak_repl2_rtsource_helper looks to have same issue in 2.0.4. Both helpers can leak.
##### leaked riak_repl2_rtsink_helper in cluster info

a parent's process looks already gone..

```
=proc:&lt;0.15572.931>
State: Waiting
Spawned as: proc_lib:init_p/5
Spawned by: &lt;0.15571.931>
Started: Sat Dec  6 15:37:53 2014
Message queue length: 0
Number of heap fragments: 0
Heap fragment data: 0
Dictionary: [{'$initial_call',{riak_repl2_rtsink_helper,init,1}},{'$ancestors',[&lt;0.15571.931>,riak_repl2_rtsink_conn_sup,riak_repl2_rtsink_sup,riak_repl2_rt_sup,riak_repl_sup,&lt;0.813.0>]}]
Reductions: 41453917
Stack+heap: 74732575
OldHeap: 0
Heap unused: 20414804
OldHeap unused: 0
Stack dump:
Program counter: 0x00007f900d944460 (gen_server:loop/6 + 264)
CP: 0x0000000000000000 (invalid)
arity = 0
```
##### setup 1 node source cluster (dev1) & 1 node sink cluster (dev2):

To understand the problem easily, we use 1 node cluster on both sides.

```
% git clone riak_ee
% cd riak_ee
% git checkout 2.0
% wget https://gist.githubusercontent.com/ksauzz/cb4bef22621e68f94c2d/raw/6e028f750366643b659291fc09051ef29ccee397/setup_repl_v3_2nodes.sh 
% sh ./setup_repl_v3_2nodes.sh 
```
##### send tcp_error to riak_repl2_rtsink_conn periodically.

simulate unstable network connectivity between source side and sink side.

```
% dev/dev2/bin/riak attach

(dev2@127.0.0.1)1> [begin hd(riak_repl2_rtsink_conn_sup:started())!{tcp_error,<<>>,test}, timer:sleep(2000) end||_<-lists:seq(1,10000)].
```
##### confirm process count at sink side.

```
% while [ true ]; do  curl -s localhost:10028/stats |python -mjson.tool | grep -e sys_proc -e memory_total ; sleep 2; done
    "memory_total": 96117504,
    "sys_process_count": 1916,
    "memory_total": 96140312,
    "sys_process_count": 1917,
    "memory_total": 96119744,
    "sys_process_count": 1918,
    "memory_total": 96121552,
    "sys_process_count": 1919,
    "memory_total": 96144016,
    "sys_process_count": 1920,
    "memory_total": 96111056,
    "sys_process_count": 1921,
    "memory_total": 96101504,
    "sys_process_count": 1922,
    "memory_total": 96124016,
    "sys_process_count": 1923,
```
##### confirm process count at source side.

```
 while [ true ]; do  curl -s localhost:10018/stats |python -mjson.tool | grep -e sys_proc -e memory_total ; sleep 2; done
    "memory_total": 96615264,
    "sys_process_count": 1967,
    "memory_total": 96655880,
    "sys_process_count": 1968,
    "memory_total": 96614616,
    "sys_process_count": 1969,
    "memory_total": 96630664,
    "sys_process_count": 1972,
    "memory_total": 96627296,
    "sys_process_count": 1971,
    "memory_total": 96626688,
    "sys_process_count": 1972,
    "memory_total": 96642608,
    "sys_process_count": 1973,
    "memory_total": 96640680,
    "sys_process_count": 1974,
    "memory_total": 96662136,
    "sys_process_count": 1975,
```

**Comment by nerophon** [comment]
This has been patched and fixed in 2.2 and 2.0.7. Please confirm.

**Comment by JeetKunDoug** [comment]
Fixed by #652 - closing.

**Comment by Basho-JIRA** [comment]
FIxed by https://github.com/basho/riak_repl/pull/652 

**_[posted via JIRA by Douglas Rohrer]**_
